apis/apps/v1alpha1/lifecycle.go

```diff
@@ -0,0 +1,40 @@
+/*
+Copyright 2020 The Kruise Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+const (
+	LifecycleStateKey     = "lifecycle.apps.kruise.io/state"
+	LifecycleTimestampKey = "lifecycle.apps.kruise.io/timestamp"
+
+	LifecycleStateNormal          LifecycleStateType = "Normal"
+	LifecycleStatePreparingUpdate LifecycleStateType = "PreparingUpdate"
+	LifecycleStateUpdating        LifecycleStateType = "Updating"
+	LifecycleStateUpdated         LifecycleStateType = "Updated"
+	LifecycleStatePreparingDelete LifecycleStateType = "PreparingDelete"
+)
+
+type LifecycleStateType string
+
+type Lifecycle struct {
+	PreDelete     *LifecycleHook `json:"preDelete,omitempty"`
+	InPlaceUpdate *LifecycleHook `json:"inPlaceUpdate,omitempty"`
+}
+
```

Comment:
I suggest to add some comments to explain how to use the hook or add a url to a github readme page later once that page is ready.